File: src/main/java/com/jj/hbase/HBaseClient.java
package com.jj.hbase;

import com.google.common.collect.Sets;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.client.Connection;
import org.apache.hadoop.hbase.client.Delete;
import org.apache.hadoop.hbase.client.Get;
import org.apache.hadoop.hbase.client.Put;
import org.apache.hadoop.hbase.client.Result;
import org.apache.hadoop.hbase.client.Table;
import org.apache.hadoop.hbase.util.Bytes;
import org.apache.hadoop.security.UserGroupInformation;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;


public class HBaseClient {
    private Connection connection;

    public HBaseClient(Connection connection) {
        this.connection = connection;
    }

    public void insert(String table, Map<String, ColumnFamily> columnFamilies) throws IOException {
        try (Table t = connection.getTable(TableName.valueOf(table))) {
            List<Put> puts = new ArrayList<>(columnFamilies.size());
            columnFamilies.forEach((rowKey, columnFamily) -> {
                final byte[] family = Bytes.toBytes(columnFamily.getName());
                Put put = new Put(Bytes.toBytes(rowKey));
                columnFamily.getColumns().forEach((column, value) -> put.addColumn(family, Bytes.toBytes(column), Bytes.toBytes(value)));
                puts.add(put);
            });

            t.put(puts);
        }
    }

    public void insert(String table, String rowKey, ColumnFamily columnFamily) throws IOException {
        try (Table t = connection.getTable(TableName.valueOf(table))) {
            final byte[] family = Bytes.toBytes(columnFamily.getName());

            Put put = new Put(Bytes.toBytes(rowKey));
            columnFamily.getColumns().forEach((column, value) -> put.addColumn(family, Bytes.toBytes(column), Bytes.toBytes(value)));

            t.put(put);
        }
    }

    public Result find(String tableName, String rowKey, String columnFamily, Set<String> columns) throws IOException {
        try (Table table = connection.getTable(TableName.valueOf(tableName))) {
            final byte[] family = Bytes.toBytes(columnFamily);
            Get get = new Get(Bytes.toBytes(rowKey));

            columns.forEach(c -> get.addColumn(family, Bytes.toBytes(c)));

            return table.get(get);
        }
    }

    public void delete(String tableName, String rowKey) throws IOException {
        try (Table table = connection.getTable(TableName.valueOf(tableName))){
            Delete delete = new Delete(Bytes.toBytes(rowKey));
            table.delete(delete);
        }
    }

    public void delete(String tableName, String rowKey, String columnFamily) throws IOException {
        try (Table table = connection.getTable(TableName.valueOf(tableName))){
            Delete delete = new Delete(Bytes.toBytes(rowKey));
            delete.addFamily(Bytes.toBytes(columnFamily));

            table.delete(delete);
        }
    }

    public static void main(String[] args) throws IOException {
        // Setting up the HBase configuration
        Configuration configuration = new Configuration();
        configuration.addResource("src/main/resources/hbase-site.xml");

        // Point to the krb5.conf file. Alternatively this could be setup when running the program using: -Djava.security.krb5.conf=<full path to krb5.conf>
        System.setProperty("java.security.krb5.conf", "src/main/resources/krb5.conf");
        System.setProperty("sun.security.krb5.debug", "true");

        String principal = System.getProperty("kerberosPrincipal", "jj@EXAMPLE.COM");
        String keytabLocation = System.getProperty("kerberosKeytab", "src/main/resources/jj.keytab");

        UserGroupInformation.setConfiguration(configuration);
        UserGroupInformation.loginUserFromKeytab(principal, keytabLocation);

        HBaseClient client = new HBaseClient(new ConnectionGetter(configuration).getConnection());
        client.insert("test", "test-java-1", new ColumnFamily("test-family").addColumn("col1", "val-java-1"));
        System.out.println(client.find("test", "test-1", "test-family", Sets.newHashSet("col1")));
    }
}

File: src/main/scala/com/jj/scala/hbase/HBaseClient.scala
package com.jj.scala.hbase

import java.io.Closeable

import org.apache.hadoop.hbase.TableName
import org.apache.hadoop.hbase.client._
import org.apache.hadoop.hbase.util.Bytes
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.hbase.HBaseConfiguration
import org.apache.hadoop.security.UserGroupInformation

import scala.util.Properties

case class HBaseTableEntry(tableName: String, rowKey: String, columnFamily: ColumnFamily)

case class HBaseTableEntries(tableName: String, rowKey: String, columnFamilies: Seq[ColumnFamily])

case class ColumnFamily(name: String, columns: Map[String, String])

class ConnectionGetter (private val conf: Configuration) {
  val connection = ConnectionFactory.createConnection(HBaseConfiguration.create(conf))
}

class HBaseClient(private val connection: Connection) {

  private def asPuts(rowKey: String, cf: ColumnFamily): Put = cf match {
    case ColumnFamily(name, columns) =>
      val family = Bytes.toBytes(name)
      val p = new Put(Bytes.toBytes(rowKey))
      columns.foreach {
        case (column, value) => p.addColumn(family, Bytes.toBytes(column), Bytes.toBytes(value))
      }
      p
  }

  def insert(e: HBaseTableEntry): Unit = e match {
    case HBaseTableEntry(tableName, rowKey, columns) =>
      TryWith(connection.getTable(TableName.valueOf(tableName))) { t =>
        val puts = asPuts(rowKey, columns)
        t.put(puts)
      }
  }

  def insert(e: HBaseTableEntries): Unit = e match {
    case HBaseTableEntries(tableName, rowKey, columnFamilies) =>
      TryWith(connection.getTable(TableName.valueOf(tableName))) { t =>
        import scala.collection.JavaConverters._

        val puts = columnFamilies.map(c => asPuts(rowKey, c))
        t.put(puts.asJava)
      }
  }

  def find(table: String, rowKey: String, familyName: String, columns: Seq[String]): Result = {
    TryWith(connection.getTable(TableName.valueOf(table))) { t =>
      val bytes = Bytes.toBytes(familyName)
      val g = new Get(Bytes.toBytes(rowKey))
      columns.foreach(c => g.addColumn(bytes, Bytes.toBytes(c)))
      t.get(g)
    }
  }

  def delete(tableName: String, rowKey: String, columnFamilies: Seq[String]): Unit = TryWith(connection.getTable(TableName.valueOf(tableName))) { t =>
    val d = new Delete(Bytes.toBytes(rowKey))
    columnFamilies.foreach { family =>
      d.addFamily(Bytes.toBytes(family))
    }
    t.delete(d)
  }

  def delete(tableName: String, rowKey: String): Unit = {
    TryWith(connection.getTable(TableName.valueOf(tableName))) { t =>
      t.delete(new Delete(Bytes.toBytes(rowKey)))
    }
  }
}

object TryWith {
  def apply[T <: Closeable, Q](c: T)(f: (T) => Q): Q = {
    try {
      f(c)
    } finally {
      c.close()
    }
  }
}

object Main extends App {
  private val configuration: Configuration = new Configuration()
  configuration.addResource("src/main/resources/hbase-site.xml")
  //configuration.addResource("src/main/resources/core-site.xml")
  //configuration.addResource("src/main/resources/hdfs-site.xml")

  System.setProperty("java.security.krb5.conf", "src/main/resources/krb5.conf")
  //System.setProperty("java.security.krb5.realm", "HORTONWORKS.LOCAL")
  //System.setProperty("java.security.krb5.kdc", "c6401.ambari.apache.org")
  //System.setProperty("sun.security.krb5.debug", "true")

  UserGroupInformation.setConfiguration(configuration)
  UserGroupInformation.loginUserFromKeytab("jj/c6401.ambari.apache.org@HORTONWORKS.LOCAL", "src/main/resources/jj.keytab")

  TryWith(new ConnectionGetter(configuration).connection) { c =>

    val client = new HBaseClient(c)
    println(client.find("test", "test-1", "test-family", Seq("col1", "val1")))
    val entry = HBaseTableEntry(
      "test",
      "test-2",
      ColumnFamily(
        "test-family",
        Map(
          "col1" -> "val1"
        )
      )
    )
    client.insert(entry)
  }
}